aspnetcore/signalr/streaming.md

```diff
@@ -13,35 +13,85 @@ uid: signalr/streaming
+## Set up the hub for streaming
+
+### Server to client streaming
+
+A hub method automatically becomes a server to client streaming hub method when it returns a `ChannelReader<T>` or a `Task<ChannelReader<T>>`. Below is a sample that shows the basics of streaming data to the client. Whenever an object is written to the `ChannelWriter` that object is immediately sent to the client. At the end, the `ChannelWriter` is completed to tell the client the stream is closed.
 
 > [!NOTE]
-> * Write to the `ChannelReader` on a background thread and return the `ChannelReader` as soon as possible. Other hub invocations will be blocked until a `ChannelReader` is returned.
+> * Write to the `ChannelWriter` on a background thread and return the `ChannelReader` as soon as possible. Other hub invocations will be blocked until a `ChannelReader` is returned.
 > * Wrap your logic in a `try ... catch` and complete the `Channel` in the catch and outside the catch to make sure the hub method invocation is completed properly.
 
 ::: moniker range="= aspnetcore-2.1"
 
-[!code-csharp[Streaming hub method](streaming/sample/Hubs/StreamHub.aspnetcore21.cs?name=snippet1)]
+[!code-csharp[Streaming hub method](streaming/sample.netcoreapp2.1/Hubs/StreamHub.cs?name=snippet1)]
+
+::: moniker-end
+
+::: moniker range="= aspnetcore-2.2"
+
+[!code-csharp[Streaming hub method](streaming/sample.netcoreapp2.2/Hubs/StreamHub.cs?name=snippet1)]
+
+::: moniker-end
+
+::: moniker range=">= aspnetcore-3.0"
+
+[!code-csharp[Streaming hub method](streaming/sample.netcoreapp3.0/Hubs/StreamHub.cs?name=snippet1)]
 
 ::: moniker-end
 
 ::: moniker range=">= aspnetcore-2.2"
 
-[!code-csharp[Streaming hub method](streaming/sample/Hubs/StreamHub.cs?name=snippet1)]
+In ASP.NET Core 2.2 or later, server to client streaming Hub methods can accept a `CancellationToken` parameter that will be triggered when the client unsubscribes from the stream. Use this token to stop the server operation and release any resources if the client disconnects before the end of the stream.
 
-In ASP.NET Core 2.2 or later, streaming Hub methods can accept a `CancellationToken` parameter that will be triggered when the client unsubscribes from the stream. Use this token to stop the server operation and release any resources if the client disconnects before the end of the stream.
+::: moniker-end
+
+::: moniker range=">= aspnetcore-3.0"
+
+### Client to server streaming
+
+A hub method automatically becomes a client to server streaming hub method when it accepts a `ChannelReader<T>`. Below is a sample that shows the basics of reading streaming data from the client. Whenever the client writes to the stream the data is written into the `ChannelReader` on the server which the hub method should be reading from.
```

Comment:
Should we link to the ChannelReader article or reference page?

Reply:
We should xref to the reference page in at least some of these so the user can get there. Usually we don't code-fence when xreffing, so it would be `accepts a [ChannelReader<T>](xref:...). Below...`

Reply:
How do xref to dotnet section of docs?!

Reply:
Paging @scottaddie, @tdykstra ? :)

Reply:
@guardrex

Reply:
https://xrefgen.azurewebsites.net/

Reply:
```
<xref:System.Threading.Channels.ChannelReader`1>
```

... if you're feel'in lucky 🍀, you can try this thing: https://xrefgen.azurewebsites.net/

... but only if your PC is paid for and a **_fried MOBO_** is something u can deal with. 😀 

i.e., that thing isn't 100% accurate. Look at the build report to see if the link doesn't work. If it isn't correct, pull up the API object/member in the API Browser, go into the `<meta>` tags and look for `ms.assetid` ... it's value should work. 🤞

Reply:
I don't think  types `<T>` work. For those use the `[](/dotnet/api/aspnetcore)` link.

@guardrex 	will correct me if I'm wrong.

Reply:
I just made updates to the XREF gen thing to compensate for `<T>`.

That app is a WIP, but it should work in *most* cases.

The ultimate source of truth is pulling up the API Browser page and looking at the `ms.assetid` meta tag value.

Absolutely *Yes!* on `/dotnet/api` links. If that's easier for engineers do that. We'll update them on a future doc update. IMO, do whatever is *fastest* 🏃.

Reply:
Worked